apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+    return Promise.resolve().then(
+      () => executeNavigationStep('afterEnter')
+    ).catch(catchStepError);
+  }
 
-    document.activeElement.blur();
+  var defer = Utils.Promise.defer();
 
-    transitionArgs.meta = {
-      next: {
-        panel: panel,
-        args: args
-      },
-      prev: currentPanel
-    };
+  function onNavigationEnd() {
+    window.removeEventListener('navigation-transition-end', onNavigationEnd);
+    window.removeEventListener('load', onNavigationEnd);
+    defer.resolve(executeNavigationStep('afterEnter'));
+  }
 
-    var nextPanelInfo = this.panels[panel];
-    var nextPanelContainer = document.getElementById(nextPanelInfo.container);
-    var nextPanelObject = this.panelObjects[nextPanelInfo.behaviour];
+  window.addEventListener('navigation-transition-end', onNavigationEnd);
+  window.addEventListener('load', onNavigationEnd); // simulate navigation end
 
-    if (!nextPanelObject) {
-      var error = new Error(
-        'Navigation.toPanel: no behaviour found for panel ' + panel
-      );
-      return Promise.reject(error);
-    }
+  return defer.promise.catch(catchStepError);
+}
 
-    var currentPanelObject;
-    var currentPanelContainer;
-    if (currentPanel) {
-      var currentPanelInfo = this.panels[currentPanel.panel];
-      currentPanelObject = this.panelObjects[currentPanelInfo.behaviour];
-      currentPanelContainer = document.getElementById(
-        currentPanelInfo.container);
-    }
+/**
+ * Used in `waitForSlideAnimation` to help solving race conditions.
+ * @type {Defer}
+ */
+var previousAnimationDefer;
+
+/**
+ * Wait for the end of an animation on `panelElement`.
+ *
+ * @param {Element} panelElement The element to watch.
+ * @returns {Promise} Resolves when the animation ends.
+ */
+function waitForSlideAnimation(panelElement) {
+  if (previousAnimationDefer) {
+    previousAnimationDefer.reject(new Error('A new animation started'));
+  }
 
-    var promise = Promise.resolve();
+  var defer = Utils.Promise.defer();
+  previousAnimationDefer = defer;
 
-    // beforeLeave and beforeEnter can return a rejected promise to prevent
-    // sliding. An exception is never expected though.
-    if (currentPanelObject && currentPanelObject.beforeLeave) {
-      promise = promise.then(
-        currentPanelObject.beforeLeave.bind(currentPanelObject, transitionArgs)
-      );
-    }
+  panelElement.addEventListener('animationend', function onAnimationEnd() {
+    this.removeEventListener('animationend', onAnimationEnd);
+    defer.resolve();
+  });
+
+  setTimeout(defer.resolve, 500);
+  return defer.promise.then(() => previousAnimationDefer = null);
+}
+
+/**
+ * Layouts the new view so that the new view's beforeEnter can deal with a
+ * rendered view.
+ */
+function prepareSwitchPanel() {
+  debug('prepareSwitchPanel()');
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
+
+  var newView = to.panel;
+
+  var shouldPrepare = !from ||
```

Comment:
I started adding this comment:
```
  /* We need to prepare the next view in 2 cases:
   * - either there is no view displayed yet (from is undefined)
   * - or we'll move to a view with the same behavior (so the same DOM element),
   *   and so the element is already displayed.
   *
   * The second case's condition should be the same than the condition for
   * doSlideAnimation in switchPanel, but the first case yields a different
   * condition.
   */
```
But then I replaced the code by

```javascript
  var noViewDisplayedYet = !from;
  var nextViewUsesSameDOMElement =
    from.panel && newView && from.panel.behavior !== newView.behavior;

  var shouldPrepare = noViewDisplayedYet || nextViewUsesSameDOMElement;
```
What do you prefer ?